modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/MappingServiceImpl.java

```diff
@@ -96,12 +108,54 @@ public CompletableFuture<List<MappedFragment>> map(MultiStepPlan multiStepPlan)
         return initialTopologyFuture.thenComposeAsync(ignore -> map0(multiStepPlan), taskExecutor);
     }
 
+    /** Called when the primary replica has expired. */
+    public CompletableFuture<Boolean> onPrimaryReplicaExpired(PrimaryReplicaEventParameters parameters) {
+        assert parameters != null;
+        assert parameters.groupId() instanceof TablePartitionId;
+
+        int tabId = ((TablePartitionId) parameters.groupId()).tableId();
+
+        return CompletableFuture.supplyAsync(() -> {
+            mappingsCache.values().removeIf(value -> value.tableIds.contains(tabId));
```

Comment:
It might make sense to wrap this in a synchronized block to reduce contention if there are will be a lot of simultaneous events about the same table (for each partition for example).